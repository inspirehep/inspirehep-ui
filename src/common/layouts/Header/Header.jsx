import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Layout, Row, Col } from 'antd';
import PropTypes from 'prop-types';
import { Set } from 'immutable';

import SearchBoxContainer from './../../containers/SearchBoxContainer';
import DropdownMenu from '../../components/DropdownMenu';
import './Header.scss';
import logo from './logo.svg';
import LoginOrUserDropdownContainer from '../../containers/LoginOrUserDropdownContainer';
import { isCataloger } from '../../authorization';
import { doSetsHaveCommonItem } from '../../utils';
import Logo from '../../../common/containers/LogoContainer';

const ONLY_SUPER_USERS_AND_CATALOGERS_AND_BETAUSERS = Set([
  'superuser',
  'cataloger',
  'betauser',
]);

const UNAUTHORIZED_TOOL_LINKS = [
  {
    href: '//inspirehep.net/textmining/',
    display: 'Reference extractor',
  },
  {
    href: '/tools/authorlist',
    display: 'Author list',
  },
  {
    href: '//inspirehep.net/info/hep/tools/bibliography_generate',
    display: 'Bibliography generator',
  },
];

const ALL_TOOL_LINKS = [
  {
    href: '/holdingpen',
    display: 'Holdingpen',
  },
  ...UNAUTHORIZED_TOOL_LINKS,
];

class Header extends Component {
  getToolLinksForUser() {
    const { userRoles } = this.props;

    if (isCataloger(userRoles)) {
      return ALL_TOOL_LINKS;
    }
    return UNAUTHORIZED_TOOL_LINKS;
  }

  displaySearchBox() {
    const { shouldDisplaySearchBox, userRoles } = this.props;
    return (
      shouldDisplaySearchBox &&
      doSetsHaveCommonItem(
        userRoles,
        ONLY_SUPER_USERS_AND_CATALOGERS_AND_BETAUSERS
      )
    );
  }

  render() {
    return (
      <Layout.Header className="__Header__">
        <Row type="flex" align="middle" gutter={16}>
          <Col lg={4} xl={5}>
            <Logo
              src={logo}
              authorizedRoles={ONLY_SUPER_USERS_AND_CATALOGERS_AND_BETAUSERS}
            />
          </Col>
          <Col lg={12} xl={13} xxl={14}>
            {this.displaySearchBox() && <SearchBoxContainer />}
          </Col>
          <Col lg={8} xl={6} xxl={5}>
            <Row type="flex" justify="end">
              <Col className="nav-item-container">
                <DropdownMenu
                  title="Tools"
                  titleClassName="nav-item"
                  items={this.getToolLinksForUser()}
                />
              </Col>
              <Col className="nav-item-container">
                <Link className="nav-item" to="/submissions/authors">
                  Submit
                </Link>
              </Col>
              <Col className="nav-item-container">
                <LoginOrUserDropdownContainer />
              </Col>
            </Row>
          </Col>
        </Row>
      </Layout.Header>
    );
  }
}
Header.propTypes = {
  shouldDisplaySearchBox: PropTypes.bool.isRequired,
  userRoles: PropTypes.instanceOf(Set).isRequired,
};

const stateToProps = state => ({
  shouldDisplaySearchBox: state.router.location.pathname !== '/',
  userRoles: Set(state.user.getIn(['data', 'roles'])),
});

export default connect(stateToProps)(Header);
