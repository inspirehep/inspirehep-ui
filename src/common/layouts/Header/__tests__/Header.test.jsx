import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import { getStoreWithState } from '../../../../fixtures/store';
import Header from '../Header';

describe('Header', () => {
  it('renders with search box if it is not on home', () => {
    const store = getStoreWithState({
      router: {
        location: {
          pathname: '/not-home',
        },
      },
    });
    const wrapper = shallow(<Header store={store} />).dive();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders without search box if it is on homepage `/`', () => {
    const store = getStoreWithState({
      router: {
        location: {
          pathname: '/',
        },
      },
    });
    const wrapper = shallow(<Header store={store} />).dive();
    expect(wrapper).toMatchSnapshot();
  });

  it('does not show authorized tool links if logged in user is not superuser nor cataloger', () => {
    const store = getStoreWithState({
      user: fromJS({
        loggedIn: true,
        data: {
          roles: ['betauser'],
        },
      }),
    });

    const wrapper = shallow(<Header store={store} />).dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('shows authorized tool links if logged in user is cataloger', () => {
    const store = getStoreWithState({
      user: fromJS({
        loggedIn: true,
        data: {
          roles: ['superuser'],
        },
      }),
    });

    const wrapper = shallow(<Header store={store} />).dive();

    expect(wrapper).toMatchSnapshot();
  });
});
