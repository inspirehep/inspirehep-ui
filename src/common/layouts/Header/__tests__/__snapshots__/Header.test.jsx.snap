// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header does not show authorized tool links if logged in user is not superuser nor cataloger 1`] = `
<Adapter
  className="__Header__"
>
  <Row
    align="middle"
    gutter={16}
    type="flex"
  >
    <Col
      lg={4}
      xl={5}
    >
      <Connect(Logo)
        authorizedRoles={
          Immutable.Set [
            "superuser",
            "cataloger",
            "betauser",
          ]
        }
        src="logo.svg"
      />
    </Col>
    <Col
      lg={12}
      xl={13}
      xxl={14}
    >
      <Connect(SearchBox) />
    </Col>
    <Col
      lg={8}
      xl={6}
      xxl={5}
    >
      <Row
        gutter={0}
        justify="end"
        type="flex"
      >
        <Col
          className="nav-item-container"
        >
          <DropdownMenu
            items={
              Array [
                Object {
                  "display": "Reference extractor",
                  "href": "//inspirehep.net/textmining/",
                },
                Object {
                  "display": "Author list",
                  "href": "/tools/authorlist",
                },
                Object {
                  "display": "Bibliography generator",
                  "href": "//inspirehep.net/info/hep/tools/bibliography_generate",
                },
              ]
            }
            title="Tools"
            titleClassName="nav-item"
          />
        </Col>
        <Col
          className="nav-item-container"
        >
          <Link
            className="nav-item"
            replace={false}
            to="/submissions/authors"
          >
            Submit
          </Link>
        </Col>
        <Col
          className="nav-item-container"
        >
          <Connect(LoginOrUserDropdownContainer) />
        </Col>
      </Row>
    </Col>
  </Row>
</Adapter>
`;

exports[`Header renders with search box if it is not on home 1`] = `
<Adapter
  className="__Header__"
>
  <Row
    align="middle"
    gutter={16}
    type="flex"
  >
    <Col
      lg={4}
      xl={5}
    >
      <Connect(Logo)
        authorizedRoles={
          Immutable.Set [
            "superuser",
            "cataloger",
            "betauser",
          ]
        }
        src="logo.svg"
      />
    </Col>
    <Col
      lg={12}
      xl={13}
      xxl={14}
    />
    <Col
      lg={8}
      xl={6}
      xxl={5}
    >
      <Row
        gutter={0}
        justify="end"
        type="flex"
      >
        <Col
          className="nav-item-container"
        >
          <DropdownMenu
            items={
              Array [
                Object {
                  "display": "Reference extractor",
                  "href": "//inspirehep.net/textmining/",
                },
                Object {
                  "display": "Author list",
                  "href": "/tools/authorlist",
                },
                Object {
                  "display": "Bibliography generator",
                  "href": "//inspirehep.net/info/hep/tools/bibliography_generate",
                },
              ]
            }
            title="Tools"
            titleClassName="nav-item"
          />
        </Col>
        <Col
          className="nav-item-container"
        >
          <Link
            className="nav-item"
            replace={false}
            to="/submissions/authors"
          >
            Submit
          </Link>
        </Col>
        <Col
          className="nav-item-container"
        >
          <Connect(LoginOrUserDropdownContainer) />
        </Col>
      </Row>
    </Col>
  </Row>
</Adapter>
`;

exports[`Header renders without search box if it is on homepage \`/\` 1`] = `
<Adapter
  className="__Header__"
>
  <Row
    align="middle"
    gutter={16}
    type="flex"
  >
    <Col
      lg={4}
      xl={5}
    >
      <Connect(Logo)
        authorizedRoles={
          Immutable.Set [
            "superuser",
            "cataloger",
            "betauser",
          ]
        }
        src="logo.svg"
      />
    </Col>
    <Col
      lg={12}
      xl={13}
      xxl={14}
    />
    <Col
      lg={8}
      xl={6}
      xxl={5}
    >
      <Row
        gutter={0}
        justify="end"
        type="flex"
      >
        <Col
          className="nav-item-container"
        >
          <DropdownMenu
            items={
              Array [
                Object {
                  "display": "Reference extractor",
                  "href": "//inspirehep.net/textmining/",
                },
                Object {
                  "display": "Author list",
                  "href": "/tools/authorlist",
                },
                Object {
                  "display": "Bibliography generator",
                  "href": "//inspirehep.net/info/hep/tools/bibliography_generate",
                },
              ]
            }
            title="Tools"
            titleClassName="nav-item"
          />
        </Col>
        <Col
          className="nav-item-container"
        >
          <Link
            className="nav-item"
            replace={false}
            to="/submissions/authors"
          >
            Submit
          </Link>
        </Col>
        <Col
          className="nav-item-container"
        >
          <Connect(LoginOrUserDropdownContainer) />
        </Col>
      </Row>
    </Col>
  </Row>
</Adapter>
`;

exports[`Header shows authorized tool links if logged in user is cataloger 1`] = `
<Adapter
  className="__Header__"
>
  <Row
    align="middle"
    gutter={16}
    type="flex"
  >
    <Col
      lg={4}
      xl={5}
    >
      <Connect(Logo)
        authorizedRoles={
          Immutable.Set [
            "superuser",
            "cataloger",
            "betauser",
          ]
        }
        src="logo.svg"
      />
    </Col>
    <Col
      lg={12}
      xl={13}
      xxl={14}
    >
      <Connect(SearchBox) />
    </Col>
    <Col
      lg={8}
      xl={6}
      xxl={5}
    >
      <Row
        gutter={0}
        justify="end"
        type="flex"
      >
        <Col
          className="nav-item-container"
        >
          <DropdownMenu
            items={
              Array [
                Object {
                  "display": "Holdingpen",
                  "href": "/holdingpen",
                },
                Object {
                  "display": "Reference extractor",
                  "href": "//inspirehep.net/textmining/",
                },
                Object {
                  "display": "Author list",
                  "href": "/tools/authorlist",
                },
                Object {
                  "display": "Bibliography generator",
                  "href": "//inspirehep.net/info/hep/tools/bibliography_generate",
                },
              ]
            }
            title="Tools"
            titleClassName="nav-item"
          />
        </Col>
        <Col
          className="nav-item-container"
        >
          <Link
            className="nav-item"
            replace={false}
            to="/submissions/authors"
          >
            Submit
          </Link>
        </Col>
        <Col
          className="nav-item-container"
        >
          <Connect(LoginOrUserDropdownContainer) />
        </Col>
      </Row>
    </Col>
  </Row>
</Adapter>
`;
